pq/lms_sign_verify.c

```diff
@@ -0,0 +1,361 @@
+        return EXIT_FAILURE;
+    }
+
+    ret = lms_sign_verify(levels, height, winternitz, sigs_to_do);
+
+    wc_FreeRng(&rng);
+
+    return (ret == 0) ? EXIT_SUCCESS : EXIT_FAILURE;
+}
+
+static void
+print_usage_and_die(void)
+{
+    fprintf(stderr, "usage:\n");
+    fprintf(stderr, "  ./lms_sign_verify <levels> <height> <winternitz> [num signatures]\n");
+    fprintf(stderr, "\n");
+    fprintf(stderr, "examples:\n");
+    fprintf(stderr, "  ./lms_sign_verify 1 5 1\n");
+    fprintf(stderr, "  ./lms_sign_verify 3 5 4 100\n");
+    fprintf(stderr, "  ./lms_sign_verify 2 10 2 0\n");
+    fprintf(stderr, "\n");
+    fprintf(stderr, "description:\n");
+    fprintf(stderr, "  Generates an LMS/HSS key pair with L=levels, H=height, and\n");
+    fprintf(stderr, "  W=winternitz parameters, then signs and verifies a given\n");
+    fprintf(stderr, "  number of signatures (1 by default).\n");
+    fprintf(stderr, "\n");
+    fprintf(stderr, "  If 0 is given for num signatures, it prints the private and\n");
+    fprintf(stderr, "  public key as hex and exits early.\n");
+    fprintf(stderr, "\n");
+    fprintf(stderr, "notes:\n");
+    fprintf(stderr, " - The acceptable parameter values are those in RFC8554:\n");
+    fprintf(stderr, "     levels = {1..8}\n");
+    fprintf(stderr, "     height = {5, 10, 15, 20, 25}\n");
+    fprintf(stderr, "     winternitz = {1, 2, 4, 8}\n");
+    fprintf(stderr, "\n");
+    fprintf(stderr, " - The number of available signatures is\n");
+    fprintf(stderr, "     n = 2 ** (levels * height)\n");
+    fprintf(stderr, "\n");
+    fprintf(stderr, " - Larger winternitz values will reduce the signature size, at\n");
+    fprintf(stderr, "   the expense of longer key generation and signature times.\n");
+    fprintf(stderr, "\n");
+    fprintf(stderr, " - Key generation time is strongly determined by the height of\n");
+    fprintf(stderr, "   the first level tree. A 3 level, 5 height tree is much faster\n");
+    fprintf(stderr, "   than 1 level, 15 height at initial key gen, even if the number\n");
+    fprintf(stderr, "   of available signatures is the same.\n");
+
+    exit(EXIT_FAILURE);
+}
+
+static bool
+write_key_file(unsigned char * priv,
+               size_t          priv_len,
+               void *          context)
+{
+    FILE *       file = NULL;
+    const char * filename = NULL;
+    size_t       n_write = 0;
+
+    if (priv == NULL || context == NULL || priv_len == 0) {
+        fprintf(stderr, "error: invalid args\n");
+        return false;
+    }
+
+    filename = context;
+
+    file = fopen(filename, "wb");
+    if (!file) {
+        fprintf(stderr, "error: fopen(%s, \"wb\") failed\n", filename);
+        return false;
+    }
+
+    n_write = fwrite(priv, 1, priv_len, file);
+
+    if (n_write != priv_len) {
+        fprintf(stderr, "error: wrote %zu, expected %zu: %d\n", n_write, priv_len,
+                ferror(file));
+        return false;
+    }
+
```

Comment:
Read back the contents into another buffer and do a memcmp() to confirm the content was written to disk.